File: services/backend/app/api/v1/endpoints/system.py
import logging
import json
from pathlib import Path
from fastapi import APIRouter
from app.services.ai.summarization_service import is_llm_working
from app.schemas.system import (
    AiStatusResponse,
    RootResponse,
    SystemStatusResponse,
)

logger = logging.getLogger(__name__)

router = APIRouter()


def _find_repo_root() -> Path:
    current = Path(__file__).resolve()
    for parent in current.parents:
        if (parent / ".git").exists() or (parent / "docs" / "incidents.json").exists():
            return parent
    # Fallback to backend root
    for parent in current.parents:
        if parent.name == "backend":
            return parent
    return current.parents[min(len(current.parents) - 1, 3)]


_REPO_ROOT = _find_repo_root()
_INCIDENTS_PATH = _REPO_ROOT / "docs" / "incidents.json"


@router.get("/", response_model=RootResponse)
def read_root():
    return {"message": "Welcome to the SkoLab API!"}


@router.get("/ai_status", response_model=AiStatusResponse)
async def ai_status():
    """Checks if the AI services have valid API keys and are reachable."""
    import os

    groq_key = os.getenv("GROQ_API")
    has_key = groq_key is not None and len(groq_key) > 10
    llm_ok = is_llm_working()
    return {
        "groq_api_configured": has_key,
        "llm_active": llm_ok,
        "model": "llama-3.3-70b-versatile",
        # Not the real key bytes: this route is unauthenticated (the admin gate
        # was retired with the metrics store, see docs/plans/2026-09-04-retire-
        # python-infra.md). Only report presence, never a prefix of the secret.
        "key_prefix": "***" if has_key else "None",
    }


@router.get("/status", response_model=SystemStatusResponse)
async def get_system_status():
    """Returns a public system status report detailing service availability and active/past incidents."""
    db_status = "operational"
    cache_status = "operational"

    # Check Database
    from app.db.database import AsyncSessionLocal
    from sqlalchemy import text

    try:
        async with AsyncSessionLocal() as session:
            await session.execute(text("SELECT 1"))
    except Exception as exc:
        logger.warning("Database health check failed: %s", exc)
        db_status = "degraded"

    # Check Cache — read-only. This is a polled status endpoint; a write per
    # call (the previous `suggestions_cache.set`) is pure overhead on the DB.
    try:
        from app.db.pg_cache import _redis_active, _redis_client

        if _redis_active and _redis_client is not None:
            await _redis_client.ping()
        else:
            # L2 is Postgres; its reachability tracks the DB probe above.
            cache_status = db_status
    except Exception as exc:
        logger.warning("Cache health check failed: %s", exc)
        cache_status = "degraded"

    # Read incidents database — resolved once at module level path
    incidents: list = []
    try:
        if _INCIDENTS_PATH.exists():
            with open(_INCIDENTS_PATH, "r", encoding="utf-8") as f:
                incidents = json.load(f)
    except Exception as exc:
        logger.error("Failed to read incidents file at %s: %s", _INCIDENTS_PATH, exc)

    active_incidents = [inc for inc in incidents if inc.get("status") != "resolved"]

    # Determine overall state — call is_llm_working() once
    llm_ok = is_llm_working()

    if db_status == "degraded" or cache_status == "degraded":
        overall_state = "outage" if db_status == "degraded" else "degraded"
    elif active_incidents:
        overall_state = "degraded"
    else:
        overall_state = "operational"

    return {
        "status": overall_state,
        "services": {
            "api_gateway": "operational",
            "database": db_status,
            "cache_layer": cache_status,
            "ai_inference": "operational" if llm_ok else "degraded",
        },
        "incidents": incidents,
    }

File: services/backend-go/internal/system/system.go
// Package system serves the non-LLM metadata routes ported from the Python
// backend's app/api/v1/endpoints/system.py: the API-router root (GET /) and the
// public status report (GET /status). GET /ai_status stays in Python — it
// reports the LLM key/health. See decisions/0010.
package system

import (
	"context"
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/skolab/backend-go/internal/db"
)

// PythonBackendURL is where the gateway reaches the Python LLM service. Mirrors
// main.go's own default; overridable in tests.
var PythonBackendURL = func() string {
	if v := os.Getenv("PYTHON_BACKEND_URL"); v != "" {
		return v
	}
	return "http://localhost:8000"
}()

// aiStatusClient is a short-timeout client for the one cross-service call /status
// makes (to Python's /api/v1/ai_status for the llm_active flag).
var aiStatusClient = &http.Client{Timeout: 2 * time.Second}

// RootResponse mirrors app/schemas/system.py::RootResponse.
type RootResponse struct {
	Message string `json:"message"`
}

// ServiceStatuses mirrors app/schemas/system.py::SystemServiceStatuses.
type ServiceStatuses struct {
	APIGateway  string `json:"api_gateway"`
	Database    string `json:"database"`
	CacheLayer  string `json:"cache_layer"`
	AIInference string `json:"ai_inference"`
}

// StatusResponse mirrors app/schemas/system.py::SystemStatusResponse.
type StatusResponse struct {
	Status    string           `json:"status"`
	Services  ServiceStatuses  `json:"services"`
	Incidents []map[string]any `json:"incidents"`
}

// Root handles GET /api/v1/ — byte-identical to the Python read_root handler.
func Root(c *gin.Context) {
	c.JSON(http.StatusOK, RootResponse{Message: "Welcome to the SkoLab API!"})
}

// Status handles GET /api/v1/status. Ports get_system_status: probe the DB and
// cache, read the incidents file, fold in the LLM-inference flag, and derive an
// overall state. The overall state depends only on DB / cache / active
// incidents — never on ai_inference — matching the Python logic exactly.
func Status(c *gin.Context) {
	ctx, cancel := context.WithTimeout(c.Request.Context(), 3*time.Second)
	defer cancel()

	dbStatus := probeDB(ctx)
	// Python: Redis L2 is PINGed when active; otherwise "the L2 is Postgres and
	// its reachability tracks the DB probe". The gateway's L2 is Postgres
	// (cache_entries), so cache_layer == database here.
	cacheStatus := dbStatus

	aiStatus := "operational"
	if !llmActive(ctx) {
		aiStatus = "degraded"
	}

	incidents := readIncidents()
	activeIncidents := 0
	for _, inc := range incidents {
		if s, _ := inc["status"].(string); s != "resolved" {
			activeIncidents++
		}
	}

	overall := "operational"
	switch {
	case dbStatus == "degraded" || cacheStatus == "degraded":
		if dbStatus == "degraded" {
			overall = "outage"
		} else {
			overall = "degraded"
		}
	case activeIncidents > 0:
		overall = "degraded"
	}

	c.JSON(http.StatusOK, StatusResponse{
		Status: overall,
		Services: ServiceStatuses{
			APIGateway:  "operational",
			Database:    dbStatus,
			CacheLayer:  cacheStatus,
			AIInference: aiStatus,
		},
		Incidents: incidents,
	})
}

func probeDB(ctx context.Context) string {
	if db.Pool == nil {
		return "degraded"
	}
	if err := db.Pool.Ping(ctx); err != nil {
		return "degraded"
	}
	return "operational"
}

// llmActive asks the Python service whether the LLM is up. Any failure to reach
// it (which is where /ai_status lives) counts as the LLM being unavailable —
// same net result as Python's is_llm_working() returning False.
func llmActive(ctx context.Context) bool {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, PythonBackendURL+"/api/v1/ai_status", nil)
	if err != nil {
		return false
	}
	resp, err := aiStatusClient.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return false
	}
	var body struct {
		LLMActive bool `json:"llm_active"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return false
	}
	return body.LLMActive
}

// readIncidents walks up from the working directory for docs/incidents.json,
// mirroring Python's _find_repo_root. In the deployed container (docker context
// is services/backend-go/) the file is absent and this returns an empty slice —
// which is exactly what the Python service also returns there, since its own
// docker context is services/backend/. A never-nil slice keeps the JSON `[]`.
func readIncidents() []map[string]any {
	out := []map[string]any{}
	dir, err := os.Getwd()
	if err != nil {
		return out
	}
	for i := 0; i < 8; i++ {
		p := filepath.Join(dir, "docs", "incidents.json")
		if raw, err := os.ReadFile(p); err == nil {
			var parsed []map[string]any
			if json.Unmarshal(raw, &parsed) == nil && parsed != nil {
				return parsed
			}
			return out
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	return out
}
